File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {

        //hashet to hold substring
        Set<Character> hset = new HashSet<>();

        int max = 0;
        int left = 0;
        int counter = 0;

        for (int right = 0; right < s.length(); right++) {

            if(!hset.contains(s.charAt(right))) {
                hset.add(s.charAt(right));
                max = Math.max(max, (right - left) + 1);
            }  else if(hset.contains(s.charAt(right))) {
                
                while(hset.contains(s.charAt(right))) {
                    hset.remove(s.charAt(left));
                    left++;
            }
            hset.add(s.charAt(right));
        }
        }
        
        return max;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.swift
class Solution {
    func lengthOfLongestSubstring(_ s: String) -> Int {

        //convert the input string into an array of char -- access
        let chars = Array(s)

        //dict initialize -- char --> most recent index seen
        var lastSeen: [Character: Int] = [:]

        //pointers
        var left = 0
        var maxLength = 0

        //iterate -- right ptr being the method of iteration
        for (right, ch) in chars.enumerated() {

            //if we see this
            if let prevIndex = lastSeen[ch], prevIndex >= left {
                left = prevIndex + 1
            }

            //update last seen
            lastSeen[ch] = right

            //updae hte max length -- right - left + 1
            let windowSize = right - left + 1
            if windowSize > maxLength {
                maxLength = windowSize
            }
        }
        return maxLength
    }
}

/*


substring - consecutive, continuous,
dont want duplicate charactets

tc - on
sc - 

dictionary -- map char - most recent

see a repeted char? move left pointer up until we remove that previous charact

two pointers - slide

*/